src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,550 @@
+        assembly_names: list[str],
+        all_assemblies: dict[str, str],
+) -> None:
+    """
+    Create standard entries in workspace and softlinks in collectionssource directory.
+
+    csd_upload: directory of assemblies that have been uploaded to workspace
+    output_dir: output directory to create entries in workspace
+    assembly_names: a list of assembly names that exist in target workspace
+    all_assemblies: a dictionary of assembly name to file path
+    """
+    assembly_ids = create_entries_in_sd_workspace(assembly_names, all_assemblies, output_dir)
+    loader_helper.create_softlinks_in_csd(csd_upload, output_dir, assembly_ids)
+
+
+def main():
+    parser = _get_parser()
+    args = parser.parse_args()
+
+    workspace_id = args.workspace_id
+    kbase_collection = getattr(args, loader_common_names.KBASE_COLLECTION_ARG_NAME)
+    source_version = getattr(args, loader_common_names.SOURCE_VER_ARG_NAME)
+    root_dir = args.root_dir
+    token_filepath = args.token_filepath
+    upload_file_ext = args.upload_file_ext
+    keep_job_dir = args.keep_job_dir
+    overwrite = args.overwrite
+
+    env = args.env
+    kb_base_url = loader_common_names.KB_BASE_URL_MAP[env]
+
+    if workspace_id <= 0:
+        parser.error(f"workspace_id needs to be > 0")
+
+    uid = os.getuid()
+    username = os.getlogin()
+
+    job_dir = loader_helper.make_job_dir(
+        root_dir, loader_common_names.SDK_JOB_DIR, username
+    )
+    csd = loader_helper.make_collection_source_dir(
+        root_dir, loader_common_names.DEFAULT_ENV, kbase_collection, source_version
+    )
+    csd_upload = loader_helper.make_collection_source_dir(
+        root_dir, loader_common_names.DEFAULT_ENV, kbase_collection, source_version, True
+    )
+    output_dir = loader_helper.make_output_dir(
+        root_dir,
+        loader_common_names.SOURCE_DATA_DIR,
+        loader_common_names.WS,
+        env,
+        workspace_id,
+    )
+
+    proc = None
+    conf = None
+
+    try:
+        # start podman service
+        proc = loader_helper.start_podman_service(uid)
+    except Exception as e:
+        raise Exception("Podman service failed to start") from e
+    else:
+        # set up conf, start callback server, and upload assemblies to workspace
+        conf = Conf(job_dir, kb_base_url, token_filepath)
+        workspace_name = conf.ws.get_workspace_info({"id": workspace_id})[1]
+
+        _, uploaded_assembly_names = _read_yaml_file(root_dir, workspace_id, env)
+        all_assemblies, wait_to_upload_assemblies = _fetch_assemblies_to_upload(
+            workspace_name,
+            csd,
+            uploaded_assembly_names,
+            upload_file_ext,
+            overwrite,
+        )
+
+        if not wait_to_upload_assemblies:
+            print(f"All {len(all_assemblies)} assembly files already exist in workspace {workspace_id}")
+            create_entries_in_ws_and_softlinks_in_csd(csd_upload, output_dir, uploaded_assembly_names, all_assemblies)
```

Comment:
If the file is already uploaded, there should already be an entry in `sourcedir/WS` right?

Reply:
@MrCreosote
right, so you suggest we just create standard entries on newly uploaded ones and assume previous assemblies already have entries in sourcedir/WS and correctly linked?

Reply:
I would say that the order of operations for an upload should be
* Upload to WS
* Write metadata file with target UPA to `/sourcedata/NCBI/None` (or the other way around I forget)

The way the script should check if something is correctly uploaded is by checking the existence / contents of the metadata file. If the file is ok there's nothing further to do other than possibly link it into the collection / source version.

If I'm missing something let me know but that seems right

Reply:
Ah, i am a little bit confused here. Your metadata file means YAML file? Or they are two different files and we need both.

Reply:
I feel like we need both

Reply:
They're both metadata files, but I'm talking about the file in the `NCBI/<genome_id>` dir, so what you're calling the YAML file

Reply:
Now I only create entries and softlink them to collections source for correctly uploaded assemblies.

Reply:
:+1: